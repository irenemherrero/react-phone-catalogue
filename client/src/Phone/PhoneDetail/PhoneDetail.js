import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import tw from 'tailwind.macro'
import { FaMobileAlt, FaMemory, FaMicrochip } from 'react-icons/fa'
import { AvailableColors } from '../../common/ColorsWrapper'

const PhoneWrapper = styled.div`
  ${tw`flex justify-center`}
`
const PhoneImage = styled.img.attrs({
  className: 'object-cover',
})`
  ${tw`h-98 w-48`}
`
//TODO: ¿generalize?
const PhoneCardInfo = styled.div`
  ${tw`flex flex-col pt-12 pl-1`}
  h1 {
    ${tw`font-semibold text-2xl`}
  }
  h2 {
    ${tw`text-gray-600`}
  }
  footer {
    ${tw`mt-4 font-semibold text-secondary text-xl`}
  }
`
const PhoneDescription = styled.div`
  ${tw`px-8 lg:px-56 xl:px-64`}
  h1 {
    ${tw`font-bold text-2xl`}
  }
  p {
    ${tw`text-justify mt-2`}
  }
`
const SpecsContainer = styled.div`
  ${tw`flex flex-col sm:flex-row`}
`
const Spec = styled.div`
  ${tw`flex flex-row mr-4 mt-4`}
  svg {
    ${tw`text-secondary h-6 w-auto`}
  }
  span {
    ${tw`text-gray-700 ml-2`}
  }
`

const PhoneDetail = (props) => {
  const {
    name,
    manufacturer,
    description,
    color,
    price,
    imageFileName,
    screen,
    processor,
    ram,
  } = props
  const urlImage = process.env.PHONES_IMAGES_URL + imageFileName
  return (
    <div>
      <PhoneWrapper>
        <PhoneImage src={urlImage} alt={name} />
        <PhoneCardInfo>
          <h1>{name}</h1>
          <h2>{manufacturer}</h2>
          <SpecsContainer>
            <Spec>
              <FaMobileAlt />
              <span>{screen}</span>
            </Spec>
            <Spec>
              <FaMicrochip />
              <span>{processor}</span>
            </Spec>
            <Spec>
              <FaMemory />
              <span>{ram} Gb</span>
            </Spec>
          </SpecsContainer>
          <AvailableColors colors={color.split(',')} />
          <footer>{price} €</footer>
        </PhoneCardInfo>
      </PhoneWrapper>
      <PhoneDescription>
        <h1>Description</h1>
        <p>{description}</p>
      </PhoneDescription>
    </div>
  )
}

PhoneDetail.propTypes = {
  name: PropTypes.string.isRequired,
  manufacturer: PropTypes.string,
  description: PropTypes.string,
  color: PropTypes.string,
  price: PropTypes.string,
  imageFileName: PropTypes.string,
  screen: PropTypes.string,
  processor: PropTypes.string,
  ram: PropTypes.string,
}
PhoneDetail.defaultProps = {
  manufactured: '',
  description: '',
  color: '',
  price: 0,
  imageFileName: 'noimage.png',
  screen: 'N/A',
  processor: 'N/A',
  ram: 'N/A',
}

export { PhoneDetail }
