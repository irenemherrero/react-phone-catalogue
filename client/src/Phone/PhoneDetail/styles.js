import styled from 'styled-components'
import tw from 'tailwind.macro'

/**
 * Wrapper content centered
 */
const DetailWrapper = styled.div`
  ${tw`flex flex-wrap justify-center h-96`}
`
/**
 * Info wrapper
 */
const PhoneCardInfo = styled.div`
  ${tw`flex flex-col pt-12 pl-1 md:pl-6`}
  h1 {
    ${tw`font-semibold text-2xl`}
  }
  h2 {
    ${tw`text-gray-600`}
  }
  footer {
    ${tw`mt-4 font-semibold text-secondary text-xl`}
  }
`
/**
 * Title with close cross
 */
const TitleClose = styled.div`
  ${tw`flex flex-row justify-between`}
  svg {
    ${tw`h-6 w-6 text-primary cursor-pointer`}
  }
`
/**
 * Description with title and paragraph
 */
const PhoneDescription = styled.div`
  ${tw`px-8 lg:px-56 xl:px-64`}
  h1 {
    ${tw`font-semibold text-2xl border-b-2 text-primary border-secondary`}
  }
  p {
    ${tw`text-justify mt-2`}
  }
`
/**
 * Responsive row/col container for specs
 */
const SpecsContainer = styled.div`
  ${tw`flex flex-col sm:flex-row`}
`
/**
 * Spec with icon
 */
const Spec = styled.div`
  ${tw`flex flex-row mr-4 mt-4`}
  svg {
    ${tw`text-secondary h-6 w-auto`}
  }
  span {
    ${tw`text-gray-700 ml-2`}
  }
`
export {
  DetailWrapper,
  PhoneCardInfo,
  TitleClose,
  PhoneDescription,
  SpecsContainer,
  Spec,
}
