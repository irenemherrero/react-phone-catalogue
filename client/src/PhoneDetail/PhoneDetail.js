import React from 'react'
import { useHistory } from 'react-router-dom'

const PhoneDetail = (props) => {
  const { imageFileName, name, price } = props
  const history = useHistory()
  // const goList = () => {
  //   history.pushState("/")
  // }
  return (
    <div>
      <div>
        {imageFileName}
        {name} - {price}
      </div>
      <button onClick={() => history.goBack()}>Volver</button>
    </div>
  )
}

export default PhoneDetail
