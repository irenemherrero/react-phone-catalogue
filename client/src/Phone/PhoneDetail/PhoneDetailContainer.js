import React from 'react'
import { useParams } from 'react-router-dom'
import { useSelector } from 'react-redux'
import { useHistory } from 'react-router-dom'

import { PhoneDetail } from './PhoneDetail'

/**
 * Functional container for phone detail
 * @component
 */
const PhoneDetailContainer = () => {
  const { id } = useParams()
  const { phones } = useSelector((state) => state)
  const history = useHistory()

  if (phones.length === 0) {
    history.push(`/`)
    return null
  }
  const selectedPhone = phones.find((item) => item.id == id)
  return <PhoneDetail {...selectedPhone} />
}

export { PhoneDetailContainer }
