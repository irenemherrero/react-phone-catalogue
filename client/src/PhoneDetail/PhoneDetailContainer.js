import React from 'react'
import { useParams } from 'react-router-dom'
import { useSelector } from 'react-redux'

import PhoneDetail from './PhoneDetail'

const PhoneDetailContainer = () => {
  const { id } = useParams()
  const { isFetching, phones } = useSelector((state) => state)
  const selectedPhone = phones.find((item) => item.id == id)
  return (
    <div>
      {isFetching ? (
        <span>loading...</span>
      ) : (
        <PhoneDetail {...selectedPhone} />
      )}
    </div>
  )
}

export { PhoneDetailContainer }
