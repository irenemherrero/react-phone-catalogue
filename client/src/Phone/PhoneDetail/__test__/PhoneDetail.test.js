import React from 'react'
import { PhoneDetail } from '../PhoneDetail'
import { render, cleanup } from '@testing-library/react'

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}))

const mockPhone = {
  name: 'iPhone 7',
  color: 'black,blue',
  manufacturer: 'Apple',
  description:
    'iPhone 7 dramatically improves the most important aspects of the iPhone experience. It introduces advanced new camera systems. The best performance and battery life ever in an iPhone. Immersive stereo speakers. The brightest, most colorful iPhone display. Splash and water resistance*. And it looks every bit as powerful as it is. This is iPhone 7.',
  price: 769,
  imageFileName: 'IPhone_7.png',
  screen: '4,7 inch IPS',
  processor: 'A10 Fusion',
  ram: 2,
}

describe('PhoneDetail', () => {
  afterEach(cleanup)

  it('render correctly', () => {
    const component = render(<PhoneDetail {...mockPhone} />)
    expect(component).toMatchSnapshot()
  })
  it('correct composition image src', () => {
    const { container } = render(<PhoneDetail {...mockPhone} />)
    expect(container.querySelector('img').getAttribute('src')).toBe(
      process.env.PHONES_IMAGES_URL + mockPhone.imageFileName
    )
  })
})
