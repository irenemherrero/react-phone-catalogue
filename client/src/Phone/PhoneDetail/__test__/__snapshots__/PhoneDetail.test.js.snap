// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneDetail render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="sc-fzoLsD bCKUm"
          data-testid="phone-detail"
        >
          <img
            alt="iPhone 7"
            class="sc-Axmtr cFtcXX object-cover"
            src="undefinedIPhone_7.png"
          />
          <div
            class="sc-fzpans KlNdT"
          >
            <div
              class="sc-fzplWN gfObLF"
            >
              <h1>
                iPhone 7
              </h1>
              <svg
                data-testid="close-detail"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 352 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                />
              </svg>
            </div>
            <h2>
              Apple
            </h2>
            <div
              class="sc-fznKkj gTLDyv"
            >
              <div
                class="sc-fznZeY cfIPFi"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M272 0H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h224c26.5 0 48-21.5 48-48V48c0-26.5-21.5-48-48-48zM160 480c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32zm112-108c0 6.6-5.4 12-12 12H60c-6.6 0-12-5.4-12-12V60c0-6.6 5.4-12 12-12h200c6.6 0 12 5.4 12 12v312z"
                  />
                </svg>
                <span>
                  4,7 inch IPS
                </span>
              </div>
              <div
                class="sc-fznZeY cfIPFi"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M416 48v416c0 26.51-21.49 48-48 48H144c-26.51 0-48-21.49-48-48V48c0-26.51 21.49-48 48-48h224c26.51 0 48 21.49 48 48zm96 58v12a6 6 0 0 1-6 6h-18v6a6 6 0 0 1-6 6h-42V88h42a6 6 0 0 1 6 6v6h18a6 6 0 0 1 6 6zm0 96v12a6 6 0 0 1-6 6h-18v6a6 6 0 0 1-6 6h-42v-48h42a6 6 0 0 1 6 6v6h18a6 6 0 0 1 6 6zm0 96v12a6 6 0 0 1-6 6h-18v6a6 6 0 0 1-6 6h-42v-48h42a6 6 0 0 1 6 6v6h18a6 6 0 0 1 6 6zm0 96v12a6 6 0 0 1-6 6h-18v6a6 6 0 0 1-6 6h-42v-48h42a6 6 0 0 1 6 6v6h18a6 6 0 0 1 6 6zM30 376h42v48H30a6 6 0 0 1-6-6v-6H6a6 6 0 0 1-6-6v-12a6 6 0 0 1 6-6h18v-6a6 6 0 0 1 6-6zm0-96h42v48H30a6 6 0 0 1-6-6v-6H6a6 6 0 0 1-6-6v-12a6 6 0 0 1 6-6h18v-6a6 6 0 0 1 6-6zm0-96h42v48H30a6 6 0 0 1-6-6v-6H6a6 6 0 0 1-6-6v-12a6 6 0 0 1 6-6h18v-6a6 6 0 0 1 6-6zm0-96h42v48H30a6 6 0 0 1-6-6v-6H6a6 6 0 0 1-6-6v-12a6 6 0 0 1 6-6h18v-6a6 6 0 0 1 6-6z"
                  />
                </svg>
                <span>
                  A10 Fusion
                </span>
              </div>
              <div
                class="sc-fznZeY cfIPFi"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 640 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M640 130.94V96c0-17.67-14.33-32-32-32H32C14.33 64 0 78.33 0 96v34.94c18.6 6.61 32 24.19 32 45.06s-13.4 38.45-32 45.06V320h640v-98.94c-18.6-6.61-32-24.19-32-45.06s13.4-38.45 32-45.06zM224 256h-64V128h64v128zm128 0h-64V128h64v128zm128 0h-64V128h64v128zM0 448h64v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h64v-96H0v96z"
                  />
                </svg>
                <span>
                  2
                   Gb
                </span>
              </div>
            </div>
            <div
              class="sc-AxhCb JKePI"
            >
              <span
                class="sc-AxhUy eyJlmj"
                color="black"
              />
              <span
                class="sc-AxhUy deRsvS"
                color="blue"
              />
            </div>
            <footer>
              769
               â‚¬
            </footer>
          </div>
        </div>
        <div
          class="sc-fznyAO fUvvwN"
        >
          <h1>
            Description
          </h1>
          <p>
            iPhone 7 dramatically improves the most important aspects of the iPhone experience. It introduces advanced new camera systems. The best performance and battery life ever in an iPhone. Immersive stereo speakers. The brightest, most colorful iPhone display. Splash and water resistance*. And it looks every bit as powerful as it is. This is iPhone 7.
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="sc-fzoLsD bCKUm"
        data-testid="phone-detail"
      >
        <img
          alt="iPhone 7"
          class="sc-Axmtr cFtcXX object-cover"
          src="undefinedIPhone_7.png"
        />
        <div
          class="sc-fzpans KlNdT"
        >
          <div
            class="sc-fzplWN gfObLF"
          >
            <h1>
              iPhone 7
            </h1>
            <svg
              data-testid="close-detail"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 352 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </div>
          <h2>
            Apple
          </h2>
          <div
            class="sc-fznKkj gTLDyv"
          >
            <div
              class="sc-fznZeY cfIPFi"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 320 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M272 0H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h224c26.5 0 48-21.5 48-48V48c0-26.5-21.5-48-48-48zM160 480c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32zm112-108c0 6.6-5.4 12-12 12H60c-6.6 0-12-5.4-12-12V60c0-6.6 5.4-12 12-12h200c6.6 0 12 5.4 12 12v312z"
                />
              </svg>
              <span>
                4,7 inch IPS
              </span>
            </div>
            <div
              class="sc-fznZeY cfIPFi"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 48v416c0 26.51-21.49 48-48 48H144c-26.51 0-48-21.49-48-48V48c0-26.51 21.49-48 48-48h224c26.51 0 48 21.49 48 48zm96 58v12a6 6 0 0 1-6 6h-18v6a6 6 0 0 1-6 6h-42V88h42a6 6 0 0 1 6 6v6h18a6 6 0 0 1 6 6zm0 96v12a6 6 0 0 1-6 6h-18v6a6 6 0 0 1-6 6h-42v-48h42a6 6 0 0 1 6 6v6h18a6 6 0 0 1 6 6zm0 96v12a6 6 0 0 1-6 6h-18v6a6 6 0 0 1-6 6h-42v-48h42a6 6 0 0 1 6 6v6h18a6 6 0 0 1 6 6zm0 96v12a6 6 0 0 1-6 6h-18v6a6 6 0 0 1-6 6h-42v-48h42a6 6 0 0 1 6 6v6h18a6 6 0 0 1 6 6zM30 376h42v48H30a6 6 0 0 1-6-6v-6H6a6 6 0 0 1-6-6v-12a6 6 0 0 1 6-6h18v-6a6 6 0 0 1 6-6zm0-96h42v48H30a6 6 0 0 1-6-6v-6H6a6 6 0 0 1-6-6v-12a6 6 0 0 1 6-6h18v-6a6 6 0 0 1 6-6zm0-96h42v48H30a6 6 0 0 1-6-6v-6H6a6 6 0 0 1-6-6v-12a6 6 0 0 1 6-6h18v-6a6 6 0 0 1 6-6zm0-96h42v48H30a6 6 0 0 1-6-6v-6H6a6 6 0 0 1-6-6v-12a6 6 0 0 1 6-6h18v-6a6 6 0 0 1 6-6z"
                />
              </svg>
              <span>
                A10 Fusion
              </span>
            </div>
            <div
              class="sc-fznZeY cfIPFi"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 640 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M640 130.94V96c0-17.67-14.33-32-32-32H32C14.33 64 0 78.33 0 96v34.94c18.6 6.61 32 24.19 32 45.06s-13.4 38.45-32 45.06V320h640v-98.94c-18.6-6.61-32-24.19-32-45.06s13.4-38.45 32-45.06zM224 256h-64V128h64v128zm128 0h-64V128h64v128zm128 0h-64V128h64v128zM0 448h64v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h64v-96H0v96z"
                />
              </svg>
              <span>
                2
                 Gb
              </span>
            </div>
          </div>
          <div
            class="sc-AxhCb JKePI"
          >
            <span
              class="sc-AxhUy eyJlmj"
              color="black"
            />
            <span
              class="sc-AxhUy deRsvS"
              color="blue"
            />
          </div>
          <footer>
            769
             â‚¬
          </footer>
        </div>
      </div>
      <div
        class="sc-fznyAO fUvvwN"
      >
        <h1>
          Description
        </h1>
        <p>
          iPhone 7 dramatically improves the most important aspects of the iPhone experience. It introduces advanced new camera systems. The best performance and battery life ever in an iPhone. Immersive stereo speakers. The brightest, most colorful iPhone display. Splash and water resistance*. And it looks every bit as powerful as it is. This is iPhone 7.
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
